Sources/Plasma/Apps/plClient/OSX/PLSView.mm

```diff
@@ -0,0 +1,286 @@
+    self.layer = self.metalLayer = layer;
+#endif
+    self.layer.backgroundColor = NSColor.blackColor.CGColor;
+    return self;
+}
+
+- (BOOL)acceptsFirstResponder
+{
+    return YES;
+}
+
+- (BOOL)wantsLayer
+{
+    return YES;
+}
+
+- (void)drawRect:(NSRect)dirtyRect {
+    [super drawRect:dirtyRect];
+}
+
+
+//MARK: Left mouse button
+-(void)mouseDown:(NSEvent *)event
+{
+    [self handleMouseButtonEvent:event];
+}
+
+-(void)mouseUp:(NSEvent *)event
+{
+    [self handleMouseButtonEvent:event];
+}
+
+-(void)mouseDragged:(NSEvent *)event
+{
+    [self updateClientMouseLocation:event];
+}
+
+//MARK: Right mouse button
+-(void)rightMouseDown:(NSEvent *)event
+{
+    [self handleMouseButtonEvent:event];
+}
+
+-(void)rightMouseUp:(NSEvent *)event
+{
+    [self handleMouseButtonEvent:event];
+}
+
+-(void)rightMouseDragged:(NSEvent *)event
+{
+    [self updateClientMouseLocation:event];
+}
+
+//MARK: Mouse movement
+-(void)mouseMoved:(NSEvent *)event {
+    [self updateClientMouseLocation:event];
+}
+
+-(void)mouseEntered:(NSEvent *)event
+{
+    [NSCursor hide];
+}
+
+-(void)mouseExited:(NSEvent *)event
+{
+    //[super mouseExited:event];
+    [NSCursor unhide];
+}
+
+//MARK: Cocoa region tracking
+-(void)updateTrackingAreas
+{
+    if(self.mouseTrackingArea) {
+        [self removeTrackingArea:self.mouseTrackingArea];
+    }
+    self.mouseTrackingArea = [[NSTrackingArea alloc] initWithRect:self.bounds options:NSTrackingMouseEnteredAndExited | NSTrackingActiveWhenFirstResponder owner:self userInfo:nil];
+    [self addTrackingArea:self.mouseTrackingArea];
+}
+
+//MARK: Mouse click handler
+-(void)handleMouseButtonEvent:(NSEvent *)event {
+    [self updateClientMouseLocation:event];
+    
+    CGPoint windowLocation = [event locationInWindow];
+    CGPoint viewLocation = [self convertPoint:windowLocation fromView:nil];
+    
+    plIMouseBEventMsg* pBMsg = new plIMouseBEventMsg;
+    
+    if(event.type == NSEventTypeLeftMouseUp) {
+        pBMsg->fButton |= kLeftButtonUp;
+    }
+    else if(event.type == NSEventTypeRightMouseUp) {
+        pBMsg->fButton |= kRightButtonUp;
+    }
+    else if(event.type == NSEventTypeLeftMouseDown) {
+        pBMsg->fButton |= kLeftButtonDown;
+    }
+    else if(event.type == NSEventTypeRightMouseDown) {
+        pBMsg->fButton |= kRightButtonDown;
+    }
+    
+    @synchronized (self.layer) {
+        self.inputManager->MsgReceive(pBMsg);
+    }
+    
+    delete(pBMsg);
+}
+
+-(void)updateClientMouseLocation:(NSEvent *)event {
+    CGPoint windowLocation = [event locationInWindow];
+    CGPoint viewLocation = [self convertPoint:windowLocation fromView:nil];
+    
+    NSRect windowViewBounds = self.bounds;
+    CGFloat deltaX = (windowLocation.x) / windowViewBounds.size.width;
+    CGFloat deltaY = (windowViewBounds.size.height - windowLocation.y) / windowViewBounds.size.height;
+    
+    plIMouseXEventMsg* pXMsg = new plIMouseXEventMsg;
+    plIMouseYEventMsg* pYMsg = new plIMouseYEventMsg;
```

Comment:
These could be local variables instead of allocated via `new`, right?

Reply:
We did that in the Windows client, but there might be some threading going on here for macOS

Reply:
This did used to be threaded on macOS - but I reverted it back to a main thread only model. There were some weird side effects going on. It could move back to a render thread model in the future though. I'll take a look at this chunk though.

Reply:
Looking at this. In plMouseDevice these are also being created as pointers, and the function for sending the mouse message expects a pointer.